Fix issue chaining multiple requests

vincendep opened the pull request:
<!--
Please open an [Issue](https://github.com/kong/insomnia/issues/new) first to discuss new
features or non-trivial changes. Please provide as much detail as possible on the change as
possible including general description, implementation details, potential shortcommings, etc.

If this PR closes an issue, please mention "Closes #XX" where #XX is the issue number.
-->
 Closes #3382

dimitropoulos commented on plugins/insomnia-plugin-response/index.js:
```diff
@@ -121,34 +121,30 @@ module.exports.templateTags = [
-          const ageSeconds = (Date.now() - response.created) / 1000;
-          shouldResend = ageSeconds > maxAgeSeconds;
-        }
+        shouldResend = ! response || (Date.now() - response.created) / 1000 > maxAgeSeconds;
```
please revert this change

dimitropoulos commented on plugins/insomnia-plugin-response/index.js:
```diff
@@ -121,34 +121,30 @@ module.exports.templateTags = [
 
       if (shouldResend && context.renderPurpose === 'send') {
         console.log('[response tag] Resending dependency');
+        const requestChain = context.context.getExtraInfo('requestChain') || []
```
Is there a reason to grab this again?  It's already declared 8 lines above (now).

vincendep replied:
nope...removed :sweat_smile:

dimitropoulos reviewed (changes requested):
Thanks for the quick PR!

Other than two very minor things, we should be ready to go with this.  I can confirm that this fixes the issue.

Here is how I tested this:
[![Run in Insomnia}](https://insomnia.rest/images/run.svg)](https://insomnia.rest/run/?label=Multiple%20Chaining%20Test&uri=https%3A%2F%2Fgist.githubusercontent.com%2Fdimitropoulos%2F64fdb4a75a5bd68998dbe1cae5cf71e4%2Fraw%2F9441f985d80e9f8e99e369689e60cb79d07b7abf%2Fmultiple-chain.json)

For this to work you'd need to run `npm run serve` from `packages/insomnia-smoke-test`.

vincendep commented:
> @vincendep sorry, I didn't catch it sooner. any idea what the changes to the package-lock are all about? I'm done for the day today but I'll take a look tomorrow and maybe we can get this merged tomorrow one way or the other.

mmm no idea what those changes are. I ve only run npm run bootstrap && npm run app-start from the root to test the changes I ve made. (But first time I was using a wrong version of node)

Commit:
Fix issue chaining multiple requests

dimitropoulos committed:
refactor: use switch to prepare for leveraging exhaustiveness checking

since this will be in TypeScript soon, and this was as topical a time as any to make this change.

dimitropoulos commented:
haha, welp.  I also nave no idea where those lockfile changes came from.  I removed them and ran a bootstrap and they didn't return.  /shrug.

All the same, I re-tested just to make sure and this seems to be working as expected.  Thanks again for contributing!

vincendep commented:
thank you too...happy to contribute!

dimitropoulos committed:
fixes failing test with new behavior (and removes unused `fromResponseTag`)

dimitropoulos commented:
ok, sorry to renege, but I would like @develohpanda to take a look at this.  There was a test that needed to be updated.  I studied it for some time, indeed, and concluded that the change we made to the behavior (to fix the bug) needed to be made to the original, but now I'm worried that there's a potential for a different use-case for recursion that there wasn't before.  The test was written in such a way that looks like it is expecting an infinite loop, but now there isn't one.  Since I'm not sure and out of an abundance of caution, let's @develohpanda a chance to take a look.

develohpanda reviewed (approved):
This looks good to me - I tested some of the recursive scenarios I could think of both on Live and in this PR and they seem to be working as intended. I like the `requestChain` approach over the boolean flag.

I think it would be good to add a safety net into the chain, however - so it doesn't grow to an unreasonable amount (maybe hard-coded at 5 or 10?) I am mostly imagining malicious use of it; if someone was to export a workspace online with many (tens, hundreds) of requests chained together, if someone was to import and run a request that would be problematic.

develohpanda commented on plugins/insomnia-plugin-response/__tests__/index.test.js:
```diff
@@ -542,19 +542,11 @@ describe('Response tag', () => {
-      try {
-        await tag.run(context, 'raw', 'req_1', '', 'always');
-      } catch (err) {
-        expect(err.message).toBe('No responses for request');
-        return;
-      }
-
-      throw new Error('Running tag should have thrown exception');
+      const response = await tag.run(context, 'raw', 'req_1', '', 'always');
+      expect(response).toBe('Response res_1')
```
@dimitropoulos I think this test needs to still test what it describes, though, this time using the `requestChain` context value instead of `fromResponseTag`

develohpanda replied:
Fixed in c4b0f25

develohpanda reviewed (changes requested):
Comment on the test which is worth addressing

develohpanda committed:
fix tests
